Fridge todo example

saicu opened the pull request:
Hi there,

I was experimenting with Tokamak and Fridge today and ended up building a small Todo API as a result.

I thought it might be a good fit as an example - let me know what you think!

Really enjoying both libraries - great work!

saicu commented:
I couldnt really get it to run without, I just wanted to partially update the fields which were sent in the payload (and in the same time, making them all optional for the req + exlude id there).

Trying to do
```
fn update(db: *fr.Session, id: u32, body: UpdateTodoReq) !void {
    return try db.update(Todo, id, body);
}
```

ends in

```
run
└─ run todos_orm_sqlite
   └─ install
      └─ install todos_orm_sqlite
         └─ zig build-exe todos_orm_sqlite Debug native 1 errors
/Users/sam/.cache/zig/p/fridge-0.9.5-3GyIikVYAQAo_R8sHa31-wARIkSRO0DQFHdnWENsvY3B/src/util.zig:144:21: error: Type mismatch for field title found:?[]const u8 expected:[]const u8
                    @compileError(
                    ^~~~~~~~~~~~~
/Users/sam/.cache/zig/p/fridge-0.9.5-3GyIikVYAQAo_R8sHa31-wARIkSRO0DQFHdnWENsvY3B/src/query.zig:139:38: note: called from here
            comptime util.checkFields(T, @TypeOf(data));
                     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
error: the following command failed with 1 compilation errors:
/opt/homebrew/Cellar/zig/0.14.0_2/bin/zig build-exe -ODebug --dep tokamak --dep fridge -Mroot=/Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/src/main.zig --dep httpz --dep embed -Mtokamak=/Users/sam/Dev/sam/zig/tokamak/src/main.zig -lsqlite3 -Mfridge=/Users/sam/.cache/zig/p/fridge-0.9.5-3GyIikVYAQAo_R8sHa31-wARIkSRO0DQFHdnWENsvY3B/src/main.zig --dep metrics --dep websocket --dep build -Mhttpz=/Users/sam/.cache/zig/p/httpz-0.0.0-PNVzrLa0BgBOHnKWUMiDXXU8iXKdc8qENAWc60QubUq6/src/httpz.zig -Membed=/Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache/c/9163eddce8373838bc5c534a5ff7d331/options.zig -Mmetrics=/Users/sam/.cache/zig/p/N-V-__8AAHOzAQBh8wB371GN1DXTl1mKs8Rdqj0sJea0U4P7/src/metrics.zig --dep build=build0 -Mwebsocket=/Users/sam/.cache/zig/p/websocket-0.1.0-ZPISdXNIAwCXG7oHBj4zc1CfmZcDeyR6hfTEOo8_YI4r/src/websocket.zig -Mbuild=/Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache/c/089178a62fda4a43cff5881531a207b8/options.zig -Mbuild0=/Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache/c/0dd4cf3a224f7a0dfefcde4635efc694/options.zig -lc --cache-dir /Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache --global-cache-dir /Users/sam/.cache/zig --name todos_orm_sqlite --zig-lib-dir /opt/homebrew/Cellar/zig/0.14.0_2/lib/zig/ --listen=- 
Build Summary: 3/8 steps succeeded; 1 failed
run transitive failure
└─ run todos_orm_sqlite transitive failure
   ├─ zig build-exe todos_orm_sqlite Debug native 1 errors
   └─ install transitive failure
      └─ install todos_orm_sqlite transitive failure
         └─ zig build-exe todos_orm_sqlite Debug native (+3 more reused dependencies)
error: the following build command failed with exit code 1:
/Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache/o/58e7ed86e930566a13efe8d03c9ba9ee/build /opt/homebrew/Cellar/zig/0.14.0_2/bin/zig /opt/homebrew/Cellar/zig/0.14.0_2/lib/zig /Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite /Users/sam/Dev/sam/zig/tokamak/examples/todos_orm_sqlite/.zig-cache /Users/sam/.cache/zig --seed 0xcddf01ff -Z8ef53cbc46345727 run
```

Any idea what I'm doing wrong? I'm still fairly new to zig, so could be me missing something.

saicu commented:
Yea you're right, I've ended up implementing it as PATCH endpoint.
I could just change the `UpdateTodoReq` fields to non optional's and just use the normal `db.update(..` if you like.

cztomsik commented:
FYI I have added support for `204 No Content` and for returning `std.http.Status`, so you don't need to inject `*Response` anymore
https://github.com/cztomsik/tokamak/commit/4048dfac0fd1b4016617e73575636735731557c6

I have few more things which would help here but they are still unfinished. Will ping you later.

cztomsik commented:
I'm going to merge this and do few changes, will ping you then so you know what and why.

saicu commented:
Alright 👍 
Just noticed I didn't add it to the ci [here](https://github.com/cztomsik/tokamak/blob/main/.github/workflows/main.yml#L30), thought it would iterate thorugh the dir.

cztomsik commented:
I've updated the example to use the new `tk.Container` feature. It's not yet documented anywhere but the idea is that instead of having to create everything yourself (in some specific order), you only define what you want to be initialized + some hooks (if necessary), and the container will figure it out itself.

On top of that, there are 2 pairs of hooks (there will be more, but these 2 will definitely stay):
- `afterBundleInit/beforeBundleDeinit` called when the container is done loading, so for example it will also run in unit-tests.
- `afterAppInit/beforeAppDeinit` will be only called if you are using `tk.app.run(&.{ App })`

> Alright 👍 Just noticed I didn't add it to the ci [here](https://github.com/cztomsik/tokamak/blob/main/.github/workflows/main.yml#L30), thought it would iterate thorugh the dir.

It did before, but I have changed that because of webview example, which would fail in github actions.

saicu commented:
> I've updated the example to use the new `tk.Container` feature. It's not yet documented anywhere but the idea is that instead of having to create everything yourself (in some specific order), you only define what you want to be initialized + some hooks (if necessary), and the container will figure it out itself.
> 
> On top of that, there are 2 pairs of hooks (there will be more, but these 2 will definitely stay):
> 
>     * `afterBundleInit/beforeBundleDeinit` called when the container is done loading, so for example it will also run in unit-tests.
> 
>     * `afterAppInit/beforeAppDeinit` will be only called if you are using `tk.app.run(&.{ App })`
> 
> 
> See [9893a59](https://github.com/cztomsik/tokamak/commit/9893a5999d1bac92ea23e00ca6985f04ecd4a686)

Nice, looks clean!

> 
> > Alright 👍 Just noticed I didn't add it to the ci [here](https://github.com/cztomsik/tokamak/blob/main/.github/workflows/main.yml#L30), thought it would iterate thorugh the dir.
> 
> It did before, but I have changed that because of webview example, which would fail in github actions.

I see, makes sense.
I've created a small pr for adding the todos to the ci [here](https://github.com/cztomsik/tokamak/pull/17)
